cmd/fleet/serve.go

```diff
@@ -385,6 +386,41 @@ the way that the Fleet server works.
 				}
 			}
 
+			if config.MDM.AppleAPNsCert != "" || config.MDM.AppleAPNsKey != "" || config.MDM.AppleSCEPCert != "" || config.MDM.AppleSCEPKey != "" {
+				// if any of the 4 flags are set and not all of them, this is an error
+				checks := map[string]string{
+					"mdm_apple_apns_cert": config.MDM.AppleAPNsCert,
+					"mdm_apple_apns_key":  config.MDM.AppleAPNsKey,
+					"mdm_apple_scep_cert": config.MDM.AppleSCEPCert,
+					"mdm_apple_scep_key":  config.MDM.AppleSCEPKey,
+				}
+				for flag, value := range checks {
+					if value == "" {
+						initFatal(fmt.Errorf("missing flag --%s", flag), "validate Apple MDM")
+					}
+				}
+
+				apnsCert, err := config.MDM.AppleAPNs()
+				if err != nil {
+					initFatal(err, "validate Apple APNs certificate and key")
+				}
+				if _, err := config.MDM.AppleSCEP(); err != nil {
+					initFatal(err, "validate Apple SCEP certificate and key")
+				}
+
+				const (
+					apnsConnectionTimeout = 10 * time.Second
+					apnsConnectionURL     = "https://api.sandbox.push.apple.com"
```

Comment:
That's the one used in Apple's documentation on how to test the APNs connection: https://developer.apple.com/documentation/usernotifications/sending_push_notifications_using_command-line_tools?language=objc

(search for `Test that you can use your certificate to connect to APNs using this command`)

Reply:
When testing using openssl, it does seem like the development vs production url matters. Using the sandbox one I get the following error

```
{"reason":"BadCertificateEnvironment"}
```

Reply:
Reading the Apple documentation a second time, the example is testing sandbox certificates, so it makes sense they instruct you to use `api.sandbox.*`.

Having said that, I've got the same as Michal but that's the response body! if you use a certificate that's not signed by Apple at all, their servers send an [alert](https://www.rfc-editor.org/rfc/rfc8446#section-6) way before you can get a response body, so for the purpose of this test it might be good enough?

Reply:
@roperzh The only mention of sandbox is in:

> A DER-encoded certificate from WWDR to connect to the APNs sandbox.

I read this as (exaggerating to make my understanding more obvious)  "you need your real, production certificate" and "for the test, we will connect to the APNs sandbox". If you follow the links from this page to get the certificate, signing request, App ID, etc. they are all links to official, production stuff. I may be wrong but I think the sandbox is the right place to ping to make this test validation (maybe they allow more requests, don't block your IP if you fail more often, etc.)

@michalnicp I think this response is ok, the validation happens before that and AIUI this is just to let you know that you're in sandbox?

I mean, I'm not 100% sure by any means, but here you can see that the cert is either "sandbox-only" or "sandbox & production": https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/establishing_a_certificate-based_connection_to_apns?language=objc

This makes me feel like it's ok to ping sandbox for a test connection.

Reply:
that sounds good to me! 👍